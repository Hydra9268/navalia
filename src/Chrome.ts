import { EventEmitter } from 'events';
import * as path from 'path';
import * as fs from 'fs';
import * as debug from 'debug';
import * as chromeUtil from './util/chrome';

const log = debug('navalia:chrome');

type triggerEvents =
  'click' |
  'mousedown' |
  'mouseup' |
  'mouseover' |
  'touchstart' |
  'touchend' |
  'focus' |
  'touchcancel' |
  'touchmove' |
  'change' |
  'blur' |
  'select'
;

export interface options {
  flags: chromeUtil.flags,
  cdp: chromeUtil.cdp | null,
}

export interface navigateOpts {
  onload?: boolean
}

export const events = {
  done: 'done',
};

export class Chrome extends EventEmitter {
  private cdp: chromeUtil.cdp | null;
  private flags: chromeUtil.flags | null;
  private kill: () => Promise<{}>;

  constructor(opts: options = { cdp: null, flags: chromeUtil.defaultFlags }) {
    super();

    this.cdp = opts.cdp;
    this.flags = opts.flags;
  }

  private async getChromeCDP(): Promise<chromeUtil.cdp> {
    if (this.cdp) {
      return this.cdp;
    }

    log(`starting chrome`);

    const { browser, cdp } = await chromeUtil.launch(this.flags || chromeUtil.defaultFlags);

    log(`chrome launched on port ${browser.port}`);

    this.kill = browser.kill;
    this.cdp = cdp;

    return cdp;
  }

  private async getSelectorId(selector: string): Promise<number | null> {
    const cdp = await this.getChromeCDP();

    const document = await cdp.DOM.getDocument();

    const { nodeId } = await cdp.DOM.querySelector({
      nodeId: document.root.nodeId,
      selector,
    });

    return nodeId;
  }

  private async trigger(eventName: triggerEvents, selector: string): Promise<any> {
    let eventClass = '';

    switch (eventName) {
      case 'click':
      case 'mousedown':
      case 'mouseup':
      case 'mouseover':
        eventClass = 'MouseEvents';
        break;

      case 'touchstart':
      case 'touchend':
      case 'touchcancel':
      case 'touchmove':
        eventClass = 'TouchEvents';
        break;

      case 'focus':
      case 'change':
      case 'blur':
      case 'select':
        eventClass = 'HTMLEvents';
        break;

      default:
        throw `chrome#trigger: Couldn't handle event ${eventName} on selector ${selector}.`;
    }

    const expression = function(selector, eventName, eventClass) {
      const node = document.querySelector(selector);
      const doc = node && node.ownerDocument ? node.ownerDocument : node;
      const e = doc && doc.createEvent(eventClass);
      e.initEvent(eventName, true);
      e.synthetic = true;
      node.dispatchEvent(e, true);
    };

    return this.evaluate(expression, selector, eventName, eventClass);
  }

  public async goto(url: string, opts: navigateOpts = { onload: true }): Promise<void> {
    const cdp = await this.getChromeCDP();

    log(`going to ${url}`);

    await cdp.Page.navigate({ url });

    if (opts.onload) {
      log(`waiting for pageload on ${url}`);
      await cdp.Page.loadEventFired();
      return;
    }

    return;
  }

  public async evaluate(expression: Function, ...args): Promise<any> {
    const cdp = await this.getChromeCDP();

    const stringifiedArgs = args.map((arg) => JSON.stringify(arg)).join(',');
    const script = `(${expression.toString()})(${stringifiedArgs})`;
    
    log(`executing script: ${script}`);
    
    return cdp.Runtime.evaluate({ expression: script, returnValue: true });
  }

  public async screenshot(filePath: string): Promise<any> {
    if (!path.isAbsolute(filePath)) {
      throw new Error(`Filepath is not absolute: ${filePath}`);
    }

    const cdp = await this.getChromeCDP();

    log(`saving screenshot to ${filePath}`);

    const base64Image = await cdp.Page.captureScreenshot();
    const buffer = new Buffer(base64Image.data, 'base64');

    return fs.writeFileSync(filePath, buffer, { encoding: 'base64' });
  }

  public async pdf(filePath: string): Promise<any> {
    if (!path.isAbsolute(filePath)) {
      throw new Error(`Filepath is not absolute: ${filePath}`);
    }
    const cdp = await this.getChromeCDP();

    log(`saving PDF to ${filePath}`);

    const base64Image = await cdp.Page.printToPDF();
    const buffer = new Buffer(base64Image.data, 'base64');

    return fs.writeFileSync(filePath, buffer, { encoding: 'base64' });
  }

  public async size(width:number, height:number): Promise<any> {
    const cdp = await this.getChromeCDP();

    log(`setting window size to ${width}x${height}`);

    return cdp.Emulation.setVisibleSize({ width, height });
  }

  public async exists(selector: string): Promise<boolean> {
    log(`checking if '${selector}' exists`);
    
    return !!await this.getSelectorId(selector);
  }

  public async html(selector: string): Promise<string | null> {
    const cdp = await this.getChromeCDP();

    log(`getting '${selector}' HTML`);

    const nodeId = await this.getSelectorId(selector);

    if (!nodeId) {
      return null;
    }

    const { outerHTML } = await cdp.DOM.getOuterHTML({ nodeId });

    return outerHTML;
  }

  public async click(selector: string): Promise<void> {
    log(`clicking '${selector}'`);

    return this.trigger('click', selector);
  }

  public async focus(selector: string): Promise<void> {
    log(`focusing '${selector}'`);

    return this.trigger('focus', selector);
  }

  public async type(selector:string, value:string): Promise<void> {
    log(`typing'${value}' into '${selector}'`);

    return this.evaluate((selector, value) => {
      var element = document.querySelector(selector);
      if (element) {
        element.value = value;
      }
    }, selector, value);
  }

  public async check(selector:string): Promise<void> {
    log(`checking checkbox '${selector}'`);

    return this.evaluate((selector) => {
      var element = document.querySelector(selector);
      if (element) {
        element.checked = true;
      }
    }, selector);
  }

  public async uncheck(selector:string): Promise<void> {
    log(`un-checking checkbox '${selector}'`);

    return this.evaluate((selector) => {
      var element = document.querySelector(selector);
      if (element) {
        element.checked = false;
      }
    }, selector);
  }

  public async select(selector: string, option:string): Promise<void> {
    log(`selecting option '${option}' in '${selector}'`);

    return this.evaluate((selector) => {
      var element = document.querySelector(selector);
      if (element) {
        element.value = option;
      }
    }, selector);
  }

  public async visible(selector: string): Promise<boolean> {
    log(`seeing if '${selector}' is visible`);

    return this.evaluate((selector) => {
      var element = document.querySelector(selector);
      if (element) {
        return (element.offsetWidth > 0 && element.offsetHeight > 0);
      }
      else return false;
    }, selector);
  }

  public async wait(time: number): Promise<object> {
    log(`waiting ${time} ms`);

    return new Promise((resolve) => { 
      setTimeout(() => resolve(), time);
    });
  }

  public done(): void {
    log(`finished`);

    if (this.kill) {
      log(`closing chrome`);
      this.kill();
    }

    this.emit(events.done);
  }
}
