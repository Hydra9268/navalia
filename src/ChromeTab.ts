import { EventEmitter } from 'events';
import * as path from 'path';
import * as fs from 'fs';
import * as debug from 'debug';

const log = debug('navalia:chrome-tab');

export type triggerEvents =
  'click' |
  'mousedown' |
  'mouseup' |
  'mouseover' |
  'touchstart' |
  'touchend' |
  'focus' |
  'touchcancel' |
  'touchmove' |
  'change' |
  'blur' |
  'select'
;

export interface navigateOpts {
  onload?: boolean
}

export const events = {
  done: 'done',
};

export class ChromeTab extends EventEmitter {
  private tab: any;
  private active: boolean;
  private targetId: string;

  constructor(tab: any, targetId: string) {
    super();
    this.active = true;
    this.tab = tab;
    this.targetId = targetId;
  }

  private async getSelectorId(selector: string): Promise<number | null> {
    const document = await this.tab.DOM.getDocument();

    const { nodeId } = await this.tab.DOM.querySelector({
      nodeId: document.root.nodeId,
      selector,
    });

    return nodeId;
  }

  private async trigger(eventName: triggerEvents, selector: string): Promise<any> {
    let eventClass = '';

    switch (eventName) {
      case 'click':
      case 'mousedown':
      case 'mouseup':
      case 'mouseover':
        eventClass = 'MouseEvents';
        break;

      case 'touchstart':
      case 'touchend':
      case 'touchcancel':
      case 'touchmove':
        eventClass = 'TouchEvents';
        break;

      case 'focus':
      case 'change':
      case 'blur':
      case 'select':
        eventClass = 'HTMLEvents';
        break;

      default:
        throw `chrome#trigger: Couldn't handle event ${eventName} on selector ${selector}.`;
    }

    const expression = function(selector, eventName, eventClass) {
      const node = document.querySelector(selector);
      const doc = node && node.ownerDocument ? node.ownerDocument : node;
      const e = doc && doc.createEvent(eventClass);
      e.initEvent(eventName, true);
      e.synthetic = true;
      node.dispatchEvent(e, true);
    };

    return this.evaluate(expression, selector, eventName, eventClass);
  }

  public async navigate(url: string, opts: navigateOpts = { onload: true }): Promise<void> {
    log(`navigating to ${url}`);
    await this.tab.Page.navigate({ url });

    if (opts.onload) {
      await this.tab.Page.loadEventFired();
      return;
    }

    return;
  }

  public async evaluate(expression: Function, ...args): Promise<any> {
    const stringifiedArgs = args.map((arg) => JSON.stringify(arg)).join(',');
    const script = `(${expression.toString()})(${stringifiedArgs})`;
    
    log(`executing script: ${script}`);
    
    return this.tab.Runtime.evaluate({ expression: script, returnValue: true });
  }

  public async screenshot(filePath: string): Promise<any> {
    if (!path.isAbsolute(filePath)) {
      throw new Error(`Filepath is not absolute: ${filePath}`);
    }
    log(`capturing screenshot ${filePath}`);

    const base64Image = await this.tab.Page.captureScreenshot();
    const buffer = new Buffer(base64Image.data, 'base64');

    return fs.writeFileSync(filePath, buffer, { encoding: 'base64' });
  }

  public async pdf(filePath: string): Promise<any> {
    if (!path.isAbsolute(filePath)) {
      throw new Error(`Filepath is not absolute: ${filePath}`);
    }
    log(`capturing PDF ${filePath}`);

    const base64Image = await this.tab.Page.printToPDF();
    const buffer = new Buffer(base64Image.data, 'base64');

    return fs.writeFileSync(filePath, buffer, { encoding: 'base64' });
  }

  public async size(width:number, height:number): Promise<any> {
    log(`setting window size ${width}x${height}`);

    return this.tab.Emulation.setVisibleSize({ width: width, height: height });
  }

  public async exists(selector: string): Promise<boolean> {
    log(`checking if '${selector}' exists`);
    
    return !!await this.getSelectorId(selector);
  }

  public async html(selector: string): Promise<string | null> {
    log(`getting '${selector}' HTML`);

    const nodeId = await this.getSelectorId(selector);

    if (!nodeId) {
      return null;
    }

    const { outerHTML } = await this.tab.DOM.getOuterHTML({ nodeId });

    return outerHTML;
  }

  public async click(selector: string): Promise<void> {
    log(`clicking '${selector}'`);
    return this.trigger('click', selector);
  }

  public async focus(selector: string): Promise<void> {
    log(`focusing '${selector}'`);
    return this.trigger('focus', selector);
  }

  public async type(selector:string, value:string) {
    log(`typing'${value}' into '${selector}'`);
    return this.evaluate((selector, value) => {
      var element = document.querySelector(selector);
      if (element) {
        element.value = value;
      }
    }, selector, value);
  }

  public async check(selector:string): Promise<void> {
    log(`checking checkbox '${selector}'`);
    return this.evaluate((selector) => {
      var element = document.querySelector(selector);
      if (element) {
        element.checked = true;
      }
    }, selector);
  }

  public async uncheck(selector:string): Promise<void> {
    log(`un-checking checkbox '${selector}'`);
    return this.evaluate((selector) => {
      var element = document.querySelector(selector);
      if (element) {
        element.checked = false;
      }
    }, selector);
  }

  public async select(selector: string, option:string): Promise<void> {
    log(`selecting option '${option}' in '${selector}'`);
    return this.evaluate((selector) => {
      var element = document.querySelector(selector);
      if (element) {
        element.value = option;
      }
    }, selector);
  }

  public async visible(selector: string): Promise<void> {
    log(`seeing if '${selector}' is visible`);
    return this.evaluate((selector) => {
      var element = document.querySelector(selector);
      if (element) {
        return (element.offsetWidth > 0 && element.offsetHeight > 0);
      }
      else return false;
    }, selector);
  }

  public async wait(time: number): Promise<object> {
    log(`waiting ${time} ms`);
   return new Promise((resolve) => { 
       setTimeout(() => resolve(), time);
   });
  }

  public done(): void {
    if (this.active) {
      log(`closing ${this.targetId}`);
      this.active = false;
      this.tab = null;
      this.emit(events.done, this.targetId);
    }
  }
}
