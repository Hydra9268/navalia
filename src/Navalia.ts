import * as debug from 'debug';

import { Chrome } from './Chrome';
import { ChromeHelper } from './util/ChromeHelper';

const log = debug('navalia');

export interface clusterParams {
  numInstances?: number
  maxJobs?: number
  workerTTL?: number
  chromeOptions?: chromeOptions
}

export interface jobFunc {
  (chromeTab: ChromeTab): Promise<any>;
}

const isBusy = (chrome: Chrome): boolean => chrome.getIsBusy();
const notBusy = (chrome: Chrome): boolean => !isBusy(chrome);

export class Navalia {
  private chromeInstances: Chrome[];
  private queueList: jobFunc[];
  private numInstances: number;
  private maxJobs: number;
  private workerTTL: number;
  private defaultChromeOptions: chromeOptions;

  constructor(opts: clusterParams = {}) {
    this.numInstances = opts.numInstances || 1;
    this.maxJobs = opts.maxJobs || -1;
    this.workerTTL = opts.workerTTL || -1;
    this.defaultChromeOptions = opts.chromeOptions || {};

    this.chromeInstances = [];
    this.queueList = [];
  }

  private destroy(chrome: Chrome): void {
    const instanceIndex = this.chromeInstances.indexOf(chrome);

    if (chrome.getIsBusy()) {
      chrome.setExpired();
      log(`instance ${chrome.port} is busy, marking for expiration when complete`);
      return;
    }

    if (instanceIndex === -1) {
      return;
    }

    // Destroy chrome instance
    chrome.quit();

    // Remove it from the instances collection
    this.chromeInstances.splice(instanceIndex, 1);

    // Launch a new on in its place
    this.launchInstance(this.defaultChromeOptions);

    log(`instance ${chrome.port} successfully closed`);
  }

  private async execute(chrome: Chrome | undefined, job: jobFunc | undefined): Promise<any> {
    if (!chrome || !job) {
      throw new Error(`#execute was called with no instance of Chrome or a Job`);
    }
    log(`instance ${chrome.port} is starting work`);

    const tab = await chrome.start();

    await job(tab);

    log(`instance ${chrome.port} has completed work`);

    tab.done();

    if (chrome.getIsExpired()) {
      log(`instance ${chrome.port} is expired and is closing`);
      this.destroy(chrome);
      return;
    }

    if (chrome.jobsComplete === this.maxJobs) {
      log(`instance ${chrome.port} has completed maximum jobs and is closing`);
      this.destroy(chrome);
      return;
    }

    if (this.queueList.length) {
      log(`instance ${chrome.port} is taking work from the queue`);
      return this.execute(chrome, this.queueList.shift());
    }

    log(`instance ${chrome.port} is idle`);
  }

  private async launchInstance(chromeOptions: chromeOptions): Promise<any> {
    const chrome = new Chrome(chromeOptions);
    await chrome.start();
    log(`instance ${chrome.port} is captured`);

    this.chromeInstances.push(chrome);

    if (this.queueList.length) {
      this.execute(chrome, this.queueList.shift());
    }

    if (this.workerTTL > 0) {
      setTimeout(() => {
        log(`instance ${chrome.port} has reached expiration`);
        this.destroy(chrome);
      }, this.workerTTL);
    }

    return chrome;
  }

  public async start(): Promise<void> {
    const startupPromise: Promise<Chrome>[] = [];

    log(`launching ${this.numInstances} instances`)

    for (let i = 0; i < this.numInstances; i++) {
      startupPromise.push(this.launchInstance(this.defaultChromeOptions));
    }

    return Promise.all(startupPromise).then(() => log(`is online and ready`));
  }

  public register(job: jobFunc): void {
    if (!this.chromeInstances.length || this.chromeInstances.every(isBusy)) {
      log('queueing work as all instances are busy');
      this.queueList.push(job);
    // Otherwise just push it through
    } else {
      log('instances are available and starting');
      this.execute(this.chromeInstances.find(notBusy), job);
    }
  }
};
