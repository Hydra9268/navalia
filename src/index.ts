import Navalia from './navalia';

export default Navalia;
