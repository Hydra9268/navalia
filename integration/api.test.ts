import { Chrome } from '../src';

describe('Chrome', () => {
  let chrome = null;

  beforeEach(() => {
    chrome = new Chrome();
  });

  afterEach(() => {
    chrome.done();
  });

  describe('#evaluate', () => {
    it('should run an expression and return the value', async () => {
      const res = await chrome.evaluate(() => window.location.href);
      expect(res).toEqual('about:blank');
    });

    it('should handle async functions', async () => {
      const res = await chrome.evaluate(() => {
        return new Promise(resolve => {
          setTimeout(() => {
            resolve(window.location.href);
          }, 10);
        });
      });
      expect(res).toEqual('about:blank');
    });

    it('should throw when errors happen in Chrome', async () => {
      return chrome
        .evaluate(() => {
          throw new Error('I should propogate');
        })
        .end()
        .then(res => {
          expect(res).toBeNull();
        })
        .catch(error => {
          expect(error).toMatchSnapshot();
        });
    });

    it('should allow variables to be passed in', async () => {
      const res = await chrome.evaluate(
        (a, b, c) => {
          return [a, b, c];
        },
        'foo',
        'bar',
        'baz',
      );
      expect(res).toEqual(['foo', 'bar', 'baz']);
    });
  });

  describe('#exists', () => {
    it('should return true if something exists', async () => {
      const exists = await chrome.exists('body');
      expect(exists).toEqual(true);
    });

    it('should return false if it does not', async () => {
      const exists = await chrome.exists('.i-am-not-there', { wait: false });
      expect(exists).toEqual(false);
    });

    it('should wait for items to appear', async () => {
      await chrome.evaluate(() => {
        setTimeout(() => {
          var div = document.createElement('div');
          div.innerHTML = '<div class="arriving-late"></div>';

          while (div.children.length > 0) {
            document.body.appendChild(div.children[0]);
          }
        }, 10);
      });
      const exists = await chrome.exists('.arriving-late');
      expect(exists).toEqual(true);
    });

    it('should wait for items to appear and return false', async () => {
      await chrome.evaluate(() => {
        setTimeout(() => {
          var div = document.createElement('div');
          div.innerHTML = '<div class="arriving-late"></div>';

          while (div.children.length > 0) {
            document.body.appendChild(div.children[0]);
          }
        }, 10);
      });
      const exists = await chrome.exists('.arriving-late', { timeout: 5 });
      expect(exists).toEqual(false);
    });
  });

  describe('#html', () => {
    it('should return the html of the page with no args', async () => {
      const html = await chrome.html();
      expect(html).toEqual('<html><head></head><body></body></html>');
    });
  });

  describe('#text', () => {
    it('should return the text an element', async () => {
      const text = `All my life I've been searching for something`;
      await chrome.evaluate(text => {
        var div = document.createElement('div');
        div.innerHTML = `<div>${text}</div>`;

        while (div.children.length > 0) {
          document.body.appendChild(div.children[0]);
        }
      }, text);
      const textRes = await chrome.text('div');
      expect(textRes).toEqual(text);
    });

    it('should wait for the element before trying', async () => {
      const text = `All my life I've been searching for something`;
      await chrome.evaluate(text => {
        setTimeout(() => {
          var div = document.createElement('div');
          div.innerHTML = `<div>${text}</div>`;

          while (div.children.length > 0) {
            document.body.appendChild(div.children[0]);
          }
        }, 5);
      }, text);
      const textRes = await chrome.text('div');
      expect(textRes).toEqual(text);
    });

    it("should throw an error if the element isn't found", async () => {
      return chrome
        .text('div', { wait: false })
        .then(textRes => {
          expect(textRes).toEqual(null);
        })
        .catch(error => {
          expect(error).toMatchSnapshot();
        });
    });
  });

  describe('#click"', () => {
    it('should click elements', async () => {
      await chrome.evaluate(() => {
        var div = document.createElement('div');
        div.innerHTML = `
          <button class="clickable" onclick="javascript:document.body.appendChild(document.createElement('span'))">Click Me!</button>
        `;

        while (div.children.length > 0) {
          document.body.appendChild(div.children[0]);
        }
      });
      await chrome.click('.clickable');
      const exists = await chrome.exists('span');
      expect(exists).toEqual(true);
    });

    it('should wait for elements by default', async () => {
      return chrome
        .evaluate(() => {
          setTimeout(() => {
            var div = document.createElement('div');
            div.innerHTML = `<button class="clickable">Click Me!</button>`;

            while (div.children.length > 0) {
              document.body.appendChild(div.children[0]);
            }
          }, 5);
        })
        .click('.clickable')
        .then(([evResult, click]) => {
          expect(click).toEqual(true);
        });
    });

    it('should throw an error for elements that are not there', async () => {
      return chrome
        .click('.nope', { wait: false })
        .then(res => expect(res).toBe(null))
        .catch(error => {
          expect(error).toMatchSnapshot();
        });
    });
  });

  describe('#focus', () => {
    it('should focus an input', () => {
      return chrome
        .evaluate(() => {
          var div = document.createElement('div');
          div.innerHTML = `<input type="text" />`;

          while (div.children.length > 0) {
            document.body.appendChild(div.children[0]);
          }
        })
        .focus('input')
        .evaluate(() => {
          return document.activeElement === document.querySelector('input');
        })
        .then(results => {
          expect(results[2]).toEqual(true);
        });
    });

    it('should wait for the element then focus', () => {
      return chrome
        .evaluate(() => {
          setTimeout(() => {
            var div = document.createElement('div');
            div.innerHTML = `<input type="text" />`;

            while (div.children.length > 0) {
              document.body.appendChild(div.children[0]);
            }
          }, 5);
        })
        .focus('input')
        .evaluate(() => {
          return document.activeElement === document.querySelector('input');
        })
        .then(results => {
          expect(results[2]).toEqual(true);
        });
    });

    it("should throw an error if the element isn't there", () => {
      return chrome
        .focus('input', { wait: false })
        .then(results => {
          expect(results).toEqual(null);
        })
        .catch(error => {
          expect(error).toMatchSnapshot();
        });
    });
  });

  describe('#type', () => {
    it('should type text into an element', () => {
      const text = 'Hello Goodbye';
      return chrome
        .evaluate(() => {
          var div = document.createElement('div');
          div.innerHTML = `<input type="text" />`;

          while (div.children.length > 0) {
            document.body.appendChild(div.children[0]);
          }
        })
        .type('input', text)
        .evaluate(() => {
          return document.querySelector('input').value;
        })
        .then(results => {
          expect(results[2]).toEqual(text);
        });
    });

    it('should wait to type text into an element', () => {
      const text = 'Hello Goodbye';
      return chrome
        .evaluate(() => {
          setTimeout(() => {
            var div = document.createElement('div');
            div.innerHTML = `<input type="text" />`;

            while (div.children.length > 0) {
              document.body.appendChild(div.children[0]);
            }
          }, 5);
        })
        .type('input', text)
        .evaluate(() => {
          return document.querySelector('input').value;
        })
        .then(results => {
          expect(results[2]).toEqual(text);
        });
    });

    it('should should throw if the element never shows up', () => {
      const text = 'Hello Goodbye';
      return chrome
        .type('input', text, { wait: false })
        .then(results => {
          expect(results).toEqual(null);
        })
        .catch(error => {
          expect(error).toMatchSnapshot();
        });
    });
  });

  describe('#check', () => {
    it('should check a checkbox', () => {
      return chrome
        .evaluate(() => {
          var div = document.createElement('div');
          div.innerHTML = `<input type="checkbox" />`;

          while (div.children.length > 0) {
            document.body.appendChild(div.children[0]);
          }
        })
        .check('input')
        .evaluate(() => {
          return document.querySelector('input').checked;
        })
        .then(results => {
          expect(results[2]).toEqual(true);
        });
    });

    it('should wait to check a checkbox', () => {
      return chrome
        .evaluate(() => {
          setTimeout(() => {
            var div = document.createElement('div');
            div.innerHTML = `<input type="checkbox" />`;

            while (div.children.length > 0) {
              document.body.appendChild(div.children[0]);
            }
          }, 5);
        })
        .check('input')
        .evaluate(() => {
          return document.querySelector('input').checked;
        })
        .then(results => {
          expect(results[2]).toEqual(true);
        });
    });

    it('should throw if the element is not found', () => {
      return chrome
        .check('input', { wait: false })
        .then(res => {
          expect(res).toEqual(null);
        })
        .catch(error => {
          expect(error).toMatchSnapshot();
        });
    });
  });

  describe('#uncheck', () => {
    it('should uncheck a checkbox', () => {
      return chrome
        .evaluate(() => {
          var div = document.createElement('div');
          div.innerHTML = `<input type="checkbox" checked/>`;

          while (div.children.length > 0) {
            document.body.appendChild(div.children[0]);
          }
        })
        .uncheck('input')
        .evaluate(() => {
          return document.querySelector('input').checked;
        })
        .then(results => {
          expect(results[2]).toEqual(false);
        });
    });

    it('should wait to check a checkbox', () => {
      return chrome
        .evaluate(() => {
          setTimeout(() => {
            var div = document.createElement('div');
            div.innerHTML = `<input type="checkbox" checked/>`;

            while (div.children.length > 0) {
              document.body.appendChild(div.children[0]);
            }
          }, 5);
        })
        .uncheck('input')
        .evaluate(() => {
          return document.querySelector('input').checked;
        })
        .then(results => {
          expect(results[2]).toEqual(false);
        });
    });

    it('should throw if the element is not found', () => {
      return chrome
        .uncheck('input', { wait: false })
        .then(results => {
          expect(results).toEqual(null);
        })
        .catch(error => {
          expect(error).toMatchSnapshot();
        });
    });
  });
});
