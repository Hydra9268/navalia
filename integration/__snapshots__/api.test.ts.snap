// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chrome #check should throw if the element is not found 1`] = `
[Error: Error: :check() > selector input doesn't exist.
    at selector (<anonymous>:9:23)
    at <anonymous>:10:16
    at <anonymous>:16:9]
`;

exports[`Chrome #click" should throw an error for elements that are not there 1`] = `
[Error: Error: :click() > Unable to find element by selector: .nope
    at click (<anonymous>:6:15)
    at <anonymous>:12:4
    at <anonymous>:18:9]
`;

exports[`Chrome #evaluate should throw when errors happen in Chrome 1`] = `
[Error: Error: I should propogate
    at result (<anonymous>:4:23)
    at <anonymous>:5:16
    at <anonymous>:11:9]
`;

exports[`Chrome #focus should throw an error if the element isn't there 1`] = `[Error: Could not find node with given id]`;

exports[`Chrome #text should throw an error if the element isn't found 1`] = `
[Error: Error: :text() > selector div wasn't found.
    at getText (<anonymous>:6:27)
    at <anonymous>:9:16
    at <anonymous>:15:9]
`;

exports[`Chrome #type should should throw if the element never shows up 1`] = `[Error: Could not find node with given id]`;

exports[`Chrome #uncheck should throw if the element is not found 1`] = `
[Error: Error: :uncheck() > Couldn't find 'input' on page.
    at selector (<anonymous>:6:27)
    at <anonymous>:10:16
    at <anonymous>:16:9]
`;
